### DISCUSSION TITLE
ART migration / generalization

### DISCUSSION THREAD

**Discussion by StefanCaldararu** [uwsbel/miniav#33] [General]
I don't think this is exactly @DanNegrut 's idea, but hopefully somewhat related. I was wondering why the idea isn't to have the abstraction at the autonomy algorithm level rather than the vehicle level. Why don't we provide a framework with a bunch of base/sample autonomous algorithms, each of which take inputs (either from sensors or from other algorithms) and produce outputs (either to other algorithms or to the vehicle). Then, the user specifies which vehicle they want to use (of which we have options - currently ART and COBRA). This is useful for someone who maybe only has their own control stack they want to test, but want a framework to integrate it into (relating to @AaronYoung5 's concern of usefulness).

This way, everything also fits in the same repo. We have this family of example autonomy algorithms, and then we also provide a few demos of example autonomy stacks (currently ART with vision, ART with GPS, and COBRA). These are full-built, but built on top of this library of functions. @AaronYoung5 in my mind, this is built in between ATK and the vehicle platform. The user uses something similar to the `atk.yml` file to specify what algorithms they want in their stack, and this gets built into the container. Is this possible or am I dreaming up something very complicated?

Also, are there other oversigthts to this idea that I am not considering? @DanNegrut is this somewhat similar to your idea or completely different?

**Comment by AaronYoung5** [comment]
But isn't there hardware specific components that wouldn't make sense as generic, ROS-level nodes? I'm unfamiliar with the code now, are the state estimation algorithms across COBRA and ART similar? What about path planning? Control would be different given the underlying vehicle models (or so I would think).

> This way, everything also fits in the same repo. We have this family of example autonomy algorithms, and then we also provide a few demos of example autonomy stacks (currently ART with vision, ART with GPS, and COBRA). These are full-built, but built on top of this library of functions. @AaronYoung5 in my mind, this is built in between ATK and the vehicle platform. The user uses something similar to the atk.yml file to specify what algorithms they want in their stack, and this gets built into the container. Is this possible or am I dreaming up something very complicated?
this would just be a launch file launching individual nodes, ya?

Overall, I like where you're going, but.... I find it hard to believe that people would be interested in the general purpose code that we write. The existing ROS ecosystem does this precisely, like there are thousands of repos/packages that are multipurpose (some input, blackbox logic, some output) and that are excessible for people to port into their projects (i.e. any driver node). I see it as overkill and time consuming to write code as if people will come and drag and drop our nodes into a working autnomy stack; if they want to do that, I would point them to ROS packages ([yolo](https://github.com/Ar-Ray-code/darknet_ros/tree/master), [state estimation](https://github.com/cra-ros-pkg/robot_localization), [control](https://control.ros.org/master/index.html)).

Do you see what I'm saying or do you disagree @StefanCaldararu?

**Reply by StefanCaldararu** [reply]
Yeah that makes sense, sorry I missed this / it's a discussion from a while ago. closing bcs I think this is also out of scope by now.
